VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,989 @@
+  ">=" as comparator (or no constraint).
+
+And ignoring all constraints with "=" or "!=" comparators, the sequence of
+constraint comparators must be an alternation of greater and lesser comparators:
+
+- "<" and "<=" must be followed by one of ">", ">=" (or no constraint).
+- ">" and ">=" must be followed by one of "<", "<=" (or no constraint).
+
+Tools must report an error for such invalid ranges.
+
+
+Parsing and validating version range specifiers
+-------------------------------------------------
+
+To parse a version range specifier string:
+
+- Remove all spaces and tabs.
+- Start from left, and split once on colon ":".
+- The left hand side is the URI-scheme that must be lowercase.
+  - Tools must validate that the URI-scheme value is ``vers``.
+- The right hand side is the specifier.
+
+- Split the specifier from left once on a slash "/".
+
+- The left hand side is the <versioning-scheme> that must be lowercase.
+  Tools should validate that the <versioning-scheme> is a known scheme.
+
+- The right hand side is a list of one or more constraints.
+  Tools must validate that this constraints string is not empty ignoring spaces.
+
+- If the constraints string is equal to "*", the ``<version-constraint>`` is "*".
+  Parsing is done and no further processing is needed for this ``vers``. A tool
+  should report an error if there are extra characters beyond "*". 
+
+- Strip leading and trailing pipes "|" from the constraints string.
+- Split the constraints on pipe "|". The result is a list of ``<version-constraint>``.
+  Consecutive pipes must be treated as one and leading and trailing pipes ignored.
+
+- For each ``<version-constraint>``:
+  - Determine if the ``<version-constraint>`` starts with one of the two comparators:
+
+    - If it starts with ">=", then the comparator is ">=".
+    - If it starts with "<=", then the comparator is "<=".
+    - If it starts with "!=", then the comparator is "!=".
+    - If it starts with "<",  then the comparator is "<".
+    - If it starts with ">",  then the comparator is ">".
+
+    - Remove the comparator from ``<version-constraint>`` string start. The
+      remaining string is the version.
+
+  - Otherwise the version is the full ``<version-constraint>`` string (which implies
+    an equality comparator of "=")
+
+  - Tools should validate and report an error if the version is empty.
+
+  - If the version contains a percent "%" character, apply URL quoting rules
+    to unquote this string.
+
+  - Append the parsed (comparator, version) to the constraints list.
+
+Finally:
+
+- The results are the ``<versioning-scheme>`` and the list of ``<comparator, version>``
+  constraints.
+
+Tools should optionally validate and simplify the list of ``<comparator, version>``
+constraints once parsing is complete:
+
+- Sort and validate the list of constraints.
+- Simplify the list of constraints.
+
+
+Version constraints simplification
```

Comment:
I think we should only allow the canonical form - simplification and transformation introduce additional places for mistakes...

Reply:
Reading [these "simplification rules"](https://github.com/package-url/purl-spec/pull/139/files/19fa4cb1683acbe1db2981b7c9dfa55369c5844e#diff-dda3c48de6da2f63bfa5902e41755b27a2f1d8c186a21dcc2641c42d4f5cfaebR472-R524), they are quite complicated.  Hence I expect this complexity to be detrimental to build a well working Purl-ecosystem.

Why is requiring implementations to output a strictly conformant list of constraints and requiring importers to error on non-conformant constraint lists deemed worse?

Reply:
From experience with PURL, implementations will get this wrong (or just skip it). Erroring on imperfect constraints instead of just accepting or fixing them would be one of the common sources of interoperability failures.

Reply:
But these "imperfect constraints" are absolutely ambiguous, hence these rules are just a wild guess what could be meant.  E.g.:
* `!=1` and `=1` is valid and to be interpreted as?  `*` would be logical, but this is not denoted.
* `<1` and `>1` is valid and to be interpreted as?  `!=1` would be logical, but this is not denoted.
* `<=1` and `!=1` is valid and to be interpreted as?  I have no idea, because they are contradicting each other.
* There are some other pairs which are contradicting, e.g. `>=1` and `!=1`.

The issue with the rules for "redundant, invalid constraints" is [well stated at the beginning of that section](https://github.com/package-url/purl-spec/pull/139/files/19fa4cb1683acbe1db2981b7c9dfa55369c5844e#diff-dda3c48de6da2f63bfa5902e41755b27a2f1d8c186a21dcc2641c42d4f5cfaebR486-R487):
```
- =, < or <= followed by < or <=: this is the same as < or <=
- > or >= followed by =, > or >=: this is the same as > or >=
```
The final `or` in each line nicely expresses "nobody can know", still the rest of this section tries to interpret them in an unambiguous fashion!?!
If I read the algorithm for discarding "redundant, invalid constraints" correctly, it simply states "for the same version number the first occurring constraint has precedence".  Why not the last one (no, this question is not meant seriously)?

Why not leaving it at "[these] redundant constrains are invalid:" (serious question)?